### PR TITLE
move qid lookup to app object, so it is updated by ecosystem import; …

### PR DESCRIPTION
**Pull request by reedstrm**
…make question/domain fetch robust to incomplete data

Turns out the way question ids are linked to book ids is not particularly robust. Also, the id-sets used to short circuit question lookup were not being updated by ecosystem import.

### CODE REVIEW

File: validator/validate_api.py [hunk 1]
```diff
@@ -72,21 +66,28 @@ def setup_parse_and_data(setup_state):
 
 def get_question_data_by_key(key, val):
```
Comment (Dantemss) [inline validator/validate_api.py:67]: Maybe add some kind of comment warning that you should only query by id-like attributes and this method will return nonsense if your query returns more than 1 question?

File: validator/validate_api.py [hunk 2]
```diff
@@ -72,21 +66,28 @@ def setup_parse_and_data(setup_state):
...
     df = current_app.df
-    first_q = df["questions"][df["questions"][key] == val].iloc[0]
-    module_id = first_q.cvuid
-    uid = first_q.uid
-    has_numeric = df["questions"][df["questions"][key] == val].iloc[0].contains_number
+    # FIXME - should use all the questions and combine associated pages
+    # FIXME - last_q works better because of some dirty data getting through
+    # that has innovation pages but not the exact book those pages are from
```
Comment (Dantemss) [inline validator/validate_api.py:71]: Hmm can you guarantee that `last_q` will be clean though?
Maybe it's better to do a select on questions with book data and possibly fallback to last if none have it?

File: validator/validate_api.py [hunk 3]
```diff
@@ -72,21 +66,28 @@ def setup_parse_and_data(setup_state):
...
+    last_q = df["questions"][df["questions"][key] == val].iloc[-1]
+    module_id = last_q.cvuid
```
Comment (Dantemss) [inline validator/validate_api.py:73]: I am interested in what logic you are using to get this.
Tutor finds all pages in the current book that matches the tags, then takes the last page in ToC order as the page the question will be assigned to.
Hopefully response-validator does the same?

**Review by Dantemss — APPROVED** [review]
Looks good, but see comments

File: validator/validate_api.py (hunk 1 above)
Comment (reedstrm) [reply validator/validate_api.py:67]: Not non-sense - it _will_ be a page that's associated with that question id, in some way, so will help. Just not as much as it should. You see now why I would like tutor to pass the book (&page) context …

File: validator/validate_api.py (hunk 2 above)
Comment (reedstrm) [reply validator/validate_api.py:71]: yes, moving the data to something that allows a join would help - and then we could control for getting the most-recent data, if the exact match isn't present.

File: validator/validate_api.py (hunk 3 above)
Comment (reedstrm) [reply validator/validate_api.py:73]: Nope, there's no inherent order  to these dataframes, as far as I can tell. Just found that last worked better than first, in the presence of some missing data. Plan to revisit this w/ an explicit ordering at some point, hence the FIXMEs :/

**Pull request merged**